adapters/madvertise/madvertise.go

```diff
@@ -0,0 +1,170 @@
+package madvertise
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpointTemplate template.Template
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpointTemplate: *template,
+	}
+
+	return bidder, nil
+}
+
+func getHeaders(request *openrtb2.BidRequest) *http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return &headers
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	placementID := ""
+	for _, imp := range request.Imp {
+		madvertiseExt, err := getImpressionExt(imp)
+		if err != nil {
+			return nil, []error{err}
+		}
+		if madvertiseExt.PlacementId != "" {
+			if placementID == "" {
+				placementID = madvertiseExt.PlacementId
+			} else if placementID != madvertiseExt.PlacementId {
+				return nil, []error{&errortypes.BadInput{
+					Message: "There must be only one placement ID",
+				}}
+			}
+		} else {
+			return nil, []error{&errortypes.BadInput{
+				Message: "The placement ID must not be empty",
+			}}
+		}
+		// request.Imp[idx].TagID = madvertiseExt.PlacementId
+	}
+
+	url, err := a.buildEndpointURL(placementID)
+	if err != nil {
+		return nil, []error{err}
+	}
+	var errors = make([]error, 0)
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		errors = append(errors, err)
+		return nil, errors
```

Comment:
Since this code path only executes if there is an error, I recommend removing the `errors` slice and using the same `[]error{err}` syntax as in other parts of this code.

Reply:
done on [master 5f1ca62e] clean error